File: library/src/main/java/com/thesurix/gesturerecycler/transactions/RevertReorderTransaction.java
package com.thesurix.gesturerecycler.transactions;

import com.thesurix.gesturerecycler.GestureAdapter;
import com.thesurix.gesturerecycler.GestureViewHolder;

/**
 * @author thesurix
 */
public class RevertReorderTransaction<T> implements AdapterTransaction {

    private final GestureAdapter<T, ? extends GestureViewHolder> mAdapter;
    private final int mFrom;
    private final int mTo;

    public RevertReorderTransaction(final GestureAdapter<T, ? extends GestureViewHolder> adapter, final int from, final int to) {
        mAdapter = adapter;
        mFrom = from;
        mTo = to;
    }

    @Override
    public boolean perform() {
        return false;
    }

    @Override
    public boolean revert() {
        final T item = mAdapter.getData().remove(mTo);
        if (item != null) {
            mAdapter.notifyItemRemoved(mTo);
            mAdapter.getData().add(mFrom, item);
            mAdapter.notifyItemInserted(mFrom);
            return true;
        }

        return false;
    }
}

File: library/src/main/java/com/thesurix/gesturerecycler/transactions/RevertReorderTransaction.kt
package com.thesurix.gesturerecycler.transactions


/**
 * @author thesurix
 */
class RevertReorderTransaction<T>(private val from: Int,
                                  private val to: Int,
                                  private val headerEnabled: Boolean) : Transaction<T> {

    override fun perform(transactional: Transactional<T>) = false

    override fun revert(transactional: Transactional<T>): Boolean {
        return with(transactional.data) {
            val item = removeAt(to)
            item?.let {
                transactional.notifyRemoved(to + if (headerEnabled) 1 else 0)
                add(from, it)
                transactional.notifyInserted(from + if (headerEnabled) 1 else 0)
                true
            } ?: false
        }
    }
}
